Week3 [STEP 2] MadCow

madcow95 opened the pull request:
@Allie-e 

# 고민했던점
- 2주차 STEP3을 진행했을 때 처럼 키보드로 입력받아 커피를 주문하는 기능을 만들었습니다.
- 잘못된 입력을 했을 때 발생할 수 있는 예외처리를 생각나는데로 만들었습니다.

# 조언을 얻고 싶은 부분
1. CoffeeOrderManager Class에서 orderCoffee() 함수 쪽에서 질문인데요!
- 커피 메뉴 dictionary에서 이름을 가져오는 코드를 보면 뭔가 빙돌아서 억지로 만들어놓은? 그런 느낌이 드는데 어떻게 생각하시는지 궁금합니다!
궁금한점 : ["아메리카노": 2000 ...] 이런식으로 만들어진 dictionary에서 키보드로 숫자를 입력받아 index를 구하고, index에 해당하는 key를 가져와 enum내부 함수를 통해
```swift
static func getCoffeeFromName(name: String) -> Coffee? {
    return self.allCases.first{ "\($0.name)" == name }
}
```
원하는 Coffee case를 가져왔습니다. 제가 만들었지만 이해가 잘 안되게 만들어진거 같아서 더 나은 방향이 있을지 조언 부탁드려요!
(뭔가 질문을 작성하는것도 어렵네요ㅠㅠ)

2. 키보드 입력에 따른 예외처리에 더 신경써야할 부분이 있을지 궁금합니다!

madcow95 committed:
CoffeeShop, Class 새로운 swift파일로 이전
Coffee enum 중복 변수명 제거

madcow95 committed:
키보드로 입력받아 커피를 선하는 기능 추가

madcow95 committed:
1. class 시작 후 예외처리 추가
2. Person, Coffee, CoffeeMenu별 swift파일 추가
3. 잔액 부족 기능 수정

madcow95 committed:
불필요 상속? 삭제

Allie-e commented on CodeStarterCamp_Week3/Coffee.swift:
```diff
@@ -0,0 +1,63 @@
+    커피를 만들면 pickUpTable 에 할당할 수 있도록 해봅시다.
+ */
+class CoffeeShop {
+    private var coffeeShopName: String
```
class 타입에 이미 CoffeeShop이라는 이름이 들어가서 name 정도로만 해도 괜찮을 것 같아요 🙂

Allie-e commented on CodeStarterCamp_Week3/Coffee.swift:
```diff
@@ -0,0 +1,63 @@
+ */
+class CoffeeShop {
+    private var coffeeShopName: String
+    private var totalSales: Int
```
이부분은 좀 더 명확하게 totalSalesAmount로 바꿔볼 수 있을 것 같습니다 ㅎㅎ

Allie-e commented on CodeStarterCamp_Week3/Coffee.swift:
```diff
@@ -0,0 +1,63 @@
+    func make(coffee: Coffee, for name: String) {
+        self.pickUpTable.append(coffee)
+        print("\(name) 님이 주문하신 \(coffee.name)(이/가) 준비되었습니다. 픽업대에서 가져가주세요.")
+        self.totalSales += coffee.price
```
for name도 좋지만 카페와 손님을 생각해봤을 때 for customer로 나타낼 수도 있을 것 같아요!

Allie-e commented on CodeStarterCamp_Week3/Coffee.swift:
```diff
@@ -0,0 +1,63 @@
+        for coffee in self.pickUpTable {
+            print(coffee.name)
+        }
+        print("-------------------------")
```
지금은 프린트 함수 안에서 구분선을 직접 넣어주고 있지만,
반복이 잦은 문자열이나 정수 같은 것들은
상수로 선언하여 관리하는 방법도 있답니다 👀
이부분은 하드코딩을 키워드로 검색 해보시면 좋을 것 같아요!
```
let separatorLine = "-------------------------"
print(separatorLine)
```

madcow95 replied:
해당 부분을 사용하지 않게 되었지만 말씀해주신 부분 공부가 되었습니다!

Allie-e commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -8,154 +8,97 @@
 }
+
+let coffeeManager: CoffeeOrderManager = CoffeeOrderManager(person: missKim, shop: yagombucks)
+coffeeManager.initializeCoffeeManager()
```
coffeeManager라는 타입 없이 요구사항을 구현하면 어떻게 할 수 있을까요?
지금 방법은 missKim이 주문을 한다기 보다는 커피매니저라는 인스턴스가 중간에서 관리를 해주는 것 같아서
요구사항에 조금 더 부합하게 변경해보면 좋을 것 같아요!

<img width="755" alt="image" src="https://github.com/yagom-academy/swift-starter-week3/assets/83864058/98a21bf8-7f3d-4d5f-ac44-373918769025">

Allie-e reviewed (commented):
광우님 이번 스텝도 고생하셨습니다~!

### 질문주신 부분
1. orderCoffee() 함수
키보드 입력 때문에 코드가 복잡해 질 수는 있을 것 같습니다.
커피 메뉴를 순서를 매겨서 출력해주려 하기 때문에 index도 필요해서 억지스럽게 느껴지셨을까요?
질문하고자 하시는 부분을 정확히 이해하기가 어렵네요ㅠㅠ..
2. 키보드 입력 예외처리
-> 이부분은 잘 해주셔서 더 신경쓸 부분은 보이지 않네요! 🙂

더 궁금하신 부분이 있다면 언제든 말씀해주세요!

madcow95 committed:
refactor : 과제 요구사항에 맞게 수정

CoffeeShop, Person 인스턴스를 중간 class에서  관리 -> Person 인스턴스에서 CoffeeShop의 메소드 호출하는 방식으로 수정

madcow95 commented:
break로 반복문을 탈출하지 않고 따로 배열에 담아 보여주는 방법으로 수정해봤습니다!
<img width="559" alt="image" src="https://github.com/yagom-academy/swift-starter-week3/assets/78129823/92dbf005-5b74-4577-9cf3-76a37e43f309">
한가지 아쉬운 점이 중복되는 커피에 대해 아메리카노 2잔 이런식으로 표현하지 못한점이 아쉽습니다..
좀 더  시간이 있을 때 공부를 더해서 구현해보도록 하겠습니다ㅠㅠ

Allie-e commented:
그 부분은 중복되는 메뉴를 카운팅하는 변수를 만드는 방법도 있을 것 같아요.
더 좋은 방법이 있을지 저도 고민해보겠습니다 ㅎㅎ
고생하셨어요!
